### PR TITLE
Topics

### PR DESCRIPTION
**Pull request by nebneb0703**
Implements the topic macro from #4 

example syntax is shown already in `network/src/topics.rs`

```rs
define_topics! {
    Coordinator {
        #[message = AuthNodeInit]
        Auth -> "coordinator/auth"
    },
    Node {
        Auth -> "node/{id}/auth"
    }
}
```

generates:

```rs
pub enum Topic {
    Coordinator(CoordinatorTopic),
    Node(NodeTopic),
}
pub struct CoordinatorAuthTopic;
impl CoordinatorAuthTopic {
    pub fn topic(&self) -> String {
        "coordinator/auth".to_owned()
    }
    pub fn message(&self) -> AuthNodeInit {
        Default::default()
    }
}
pub enum CoordinatorTopic {
    Auth,
}
pub struct NodeAuthTopic;
impl NodeAuthTopic {
    pub fn topic(&self, id: String) -> String {
        format!("node/{id}/auth")
    }
}
pub enum NodeTopic {
    Auth,
}

macro_rules! topic {
    (Coordinator::Auth) => { CoordinatorAuthTopic };
    (Node::Auth) => { NodeAuthTopic };
}
```

### CODE REVIEW

**Comment by nebneb0703** [comment]
should probably add some docs to the macro before merging

File: macros/Cargo.toml [hunk 1]
```diff
@@ -10,4 +10,7 @@ proc-macro = true
 
 [dependencies]
 syn = "1.0.95"
-quote = "1.0.18"
\ No newline at end of file
+quote = "1.0.18"
+proc-macro2 = "1.0.39"
+regex = "1.5.6"
+lazy_static = "1.4.0"
```
Comment (profi248) [inline macros/Cargo.toml:16]: please add trailing newlines

File: macros/src/topics.rs [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+    }]);
+
+    enum_tokens.extend([quote! {
+        #variant_name,
+    }]);
+
+    macro_tokens.extend([quote! {
+        (#macro_path) => { #struct_name };
+    }]);
+}
```
Comment (profi248) [inline macros/src/topics.rs:260]: please add a trailing newline here as well

File: network/src/topics.rs [hunk 3]
```diff
@@ -1,49 +1,13 @@
...
-        .into()
+define_topics! {
+    Coordinator {
+        #[message = AuthNodeInit]
+        Auth -> "coordinator/auth"
+    },
+    Node {
+        Auth -> "node/{id}/auth"
     }
 }
```
Comment (profi248) [inline network/src/topics.rs:13]: here is one missing too

**Review by profi248 — CHANGES_REQUESTED** [review]

File: macros/Cargo.toml (hunk 1 above)
Comment (nebneb0703) [reply macros/Cargo.toml:16]: 😭

File: macros/src/topics.rs (hunk 2 above)
Comment (nebneb0703) [reply macros/src/topics.rs:260]: 😭

File: network/src/topics.rs (hunk 3 above)
Comment (nebneb0703) [reply network/src/topics.rs:13]: 😭

**Comment by profi248** [comment]
very nice~

**Pull request merged**